share/p2p/shrexnd/server.go

```diff
@@ -0,0 +1,168 @@
+package shrexnd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/minio/sha256-simd"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/getters"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/shrexnd/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+)
+
+const ndProtocolID = "/shrex/nd/0.0.1"
+
+var log = logging.Logger("shrex/nd")
+
+const (
+	writeTimeout = time.Second * 10
+	readTimeout  = time.Second * 10
+	serveTimeout = time.Second * 10
+)
+
+type Server struct {
+	getter       share.Getter
+	store        *eds.Store
+	host         host.Host
+	writeTimeout time.Duration
+	readTimeout  time.Duration
+	serveTimeout time.Duration
+}
+
+func StartServer(host host.Host, store *eds.Store, getter *getters.IPLDGetter) *Server {
+	srv := &Server{
+		getter:       getter,
+		store:        store,
+		host:         host,
+		writeTimeout: writeTimeout,
+		readTimeout:  readTimeout,
+		serveTimeout: serveTimeout,
+	}
+	host.SetStreamHandler(ndProtocolID, srv.handleNamespacedData)
+	return srv
+}
+
+// Stop stops the server
+func (srv *Server) Stop() {
+	srv.host.RemoveStreamHandler(ndProtocolID)
+}
+
+func (srv *Server) handleNamespacedData(stream network.Stream) {
+	err := stream.SetReadDeadline(time.Now().Add(srv.readTimeout))
+	if err != nil {
+		log.Debugf("server: seting read deadline: %s", err)
+	}
+
+	var req share_p2p_v1.GetSharesByNamespaceRequest
+	_, err = serde.Read(stream, &req)
+	if err != nil {
+		log.Errorw("server: reading request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	err = stream.CloseRead()
+	if err != nil {
+		log.Debugf("server: closing read side of the stream: %s", err)
+	}
+
+	err = validateRequest(req)
+	if err != nil {
+		log.Errorw("server: invalid request", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), srv.serveTimeout)
```

Comment:
maybe we should store a ctx, cancel on the struct during `Start` (and `cancel`s on `Stop`) and use it here instead of Background

Reply:
I've passed contexted directly, and saved `cancel` for `Stop` call